Exception thrown when using private class method

### Steps to reproduce

Given a file `test.py`:
```python
"""Bla"""

class Bla:
    """Bla"""
    def __init__(self):
        type(self).__a()

    @classmethod
    def __a(cls):
        pass

```

### Current behavior
Result of `pylint a.py`:
```
C:\temp>py -3 -m pylint test.py
Exception on node <ClassDef.Bla l.3 at 0x1255a3b6dc0> in file 'C:\temp\test.py'
Traceback (most recent call last):
  File "C:\Program Files\Python39\lib\runpy.py", line 197, in _run_module_as_main
    return _run_code(code, main_globals, None,
  File "C:\Program Files\Python39\lib\runpy.py", line 87, in _run_code
    exec(code, run_globals)
  File "C:\Program Files\Python39\lib\site-packages\pylint\__main__.py", line 9, in <module>
    pylint.run_pylint()
  File "C:\Program Files\Python39\lib\site-packages\pylint\__init__.py", line 24, in run_pylint
    PylintRun(sys.argv[1:])
  File "C:\Program Files\Python39\lib\site-packages\pylint\lint\run.py", line 384, in __init__
    linter.check(args)
  File "C:\Program Files\Python39\lib\site-packages\pylint\lint\pylinter.py", line 973, in check
    self._check_files(
  File "C:\Program Files\Python39\lib\site-packages\pylint\lint\pylinter.py", line 1007, in _check_files
    self._check_file(get_ast, check_astroid_module, name, filepath, modname)
  File "C:\Program Files\Python39\lib\site-packages\pylint\lint\pylinter.py", line 1033, in _check_file
    check_astroid_module(ast_node)
  File "C:\Program Files\Python39\lib\site-packages\pylint\lint\pylinter.py", line 1170, in check_astroid_module
    retval = self._check_astroid_module(
  File "C:\Program Files\Python39\lib\site-packages\pylint\lint\pylinter.py", line 1215, in _check_astroid_module
    walker.walk(ast_node)
  File "C:\Program Files\Python39\lib\site-packages\pylint\utils\ast_walker.py", line 77, in walk
    self.walk(child)
  File "C:\Program Files\Python39\lib\site-packages\pylint\utils\ast_walker.py", line 79, in walk
    callback(astroid)
  File "C:\Program Files\Python39\lib\site-packages\pylint\checkers\classes.py", line 901, in leave_classdef
    self._check_unused_private_members(node)
  File "C:\Program Files\Python39\lib\site-packages\pylint\checkers\classes.py", line 914, in _check_unused_private_members
    and attribute.expr.name in ("self", node.name)
AttributeError: 'Call' object has no attribute 'name'
```

### Expected behavior

```
************* Module test
test.py:3:0: R0903: Too few public methods (0/2) (too-few-public-methods)

------------------------------------------------------------------
Your code has been rated at 8.00/10 (previous run: 2.00/10, +6.00)
```

### pylint --version output

Result of `pylint --version` output:

```
C:\temp>py -3 -m pylint --version
pylint 2.9.1
astroid 2.6.2
Python 3.9.5 (tags/v3.9.5:0a7dcbd, May  3 2021, 17:27:52) [MSC v.1928 64 bit (AMD64)]
```

The issue doesn't reproduce on version 2.8.3
It started reproducing on 2.9.0 and still happens in the latest prerelease version

--- a/ChangeLog
+++ b/ChangeLog
@@ -17,6 +17,10 @@ Release date: TBA
 ..
   Put bug fixes that should not wait for a new minor version here
 
+* Fix a crash that happened when analysing code using ``type(self)`` to access
+  a class attribute in the ``unused-private-member`` checker.
+
+  Closes #4638
 
 
 What's New in Pylint 2.9.1?

--- a/pylint/checkers/classes.py
+++ b/pylint/checkers/classes.py
@@ -905,7 +905,6 @@ a metaclass class method.",
         # Check for unused private functions
         for function_def in node.nodes_of_class(astroid.FunctionDef):
             function_def = cast(astroid.FunctionDef, function_def)
-            found = False
             if not is_attr_private(function_def.name):
                 continue
             for attribute in node.nodes_of_class(astroid.Attribute):
@@ -915,16 +914,13 @@ a metaclass class method.",
                     and attribute.scope() != function_def  # We ignore recursive calls
                     and attribute.expr.name in ("self", node.name)
                 ):
-                    found = True
                     break
-            if not found:
+            else:
+                function_repr = f"{function_def.name}({function_def.args.as_string()})"
                 self.add_message(
                     "unused-private-member",
                     node=function_def,
-                    args=(
-                        node.name,
-                        f"{function_def.name}({function_def.args.as_string()})",
-                    ),
+                    args=(node.name, function_repr),
                 )
 
         # Check for unused private variables

--- a/pylint/checkers/classes.py
+++ b/pylint/checkers/classes.py
@@ -46,7 +46,7 @@
 """
 import collections
 from itertools import chain, zip_longest
-from typing import List, Pattern
+from typing import List, Pattern, cast
 
 import astroid
 
@@ -904,10 +904,12 @@ a metaclass class method.",
     def _check_unused_private_members(self, node: astroid.ClassDef) -> None:
         # Check for unused private functions
         for function_def in node.nodes_of_class(astroid.FunctionDef):
+            function_def = cast(astroid.FunctionDef, function_def)
             found = False
             if not is_attr_private(function_def.name):
                 continue
             for attribute in node.nodes_of_class(astroid.Attribute):
+                attribute = cast(astroid.Attribute, attribute)
                 if (
                     attribute.attrname == function_def.name
                     and attribute.scope() != function_def  # We ignore recursive calls

--- a/pylint/checkers/classes.py
+++ b/pylint/checkers/classes.py
@@ -909,14 +909,29 @@ a metaclass class method.",
                 continue
             for attribute in node.nodes_of_class(astroid.Attribute):
                 attribute = cast(astroid.Attribute, attribute)
-                if attribute.expr is None:  # attribute.expr is probably not really an Optional ?
+                if attribute.expr is None:
+                    # attribute.expr is an Optional[NodeNG] it can be None
                     continue
                 if (
-                    attribute.attrname == function_def.name
-                    and attribute.scope() != function_def  # We ignore recursive calls
-                    and attribute.expr.name in ("self", node.name)
+                    attribute.attrname != function_def.name
+                    or attribute.scope() == function_def
                 ):
+                    # We ignore recursive calls
+                    continue
+                if isinstance(attribute.expr, astroid.Name) and attribute.expr.name in (
+                    "self",
+                    node.name,
+                ):
+                    # self.__attrname / node_name.__attrname
                     break
+                if isinstance(attribute.expr, astroid.Call):
+                    # type(self).__attrname
+                    inferred = safe_infer(attribute.expr)
+                    if (
+                        isinstance(inferred, astroid.ClassDef)
+                        and inferred.name == node.name
+                    ):
+                        break
             else:
                 function_repr = f"{function_def.name}({function_def.args.as_string()})"
                 self.add_message(

--- a/pylint/checkers/classes.py
+++ b/pylint/checkers/classes.py
@@ -909,6 +909,8 @@ a metaclass class method.",
                 continue
             for attribute in node.nodes_of_class(astroid.Attribute):
                 attribute = cast(astroid.Attribute, attribute)
+                if attribute.expr is None:  # attribute.expr is probably not really an Optional ?
+                    continue
                 if (
                     attribute.attrname == function_def.name
                     and attribute.scope() != function_def  # We ignore recursive calls

--- a/pylint/checkers/classes.py
+++ b/pylint/checkers/classes.py
@@ -909,14 +909,10 @@ a metaclass class method.",
                 continue
             for attribute in node.nodes_of_class(astroid.Attribute):
                 attribute = cast(astroid.Attribute, attribute)
-                if attribute.expr is None:
-                    # attribute.expr is an Optional[NodeNG] it can be None
-                    continue
                 if (
                     attribute.attrname != function_def.name
-                    or attribute.scope() == function_def
+                    or attribute.scope() == function_def  # We ignore recursive calls
                 ):
-                    # We ignore recursive calls
                     continue
                 if isinstance(attribute.expr, astroid.Name) and attribute.expr.name in (
                     "self",

--- a/tests/functional/u/unused/unused_private_member.py
+++ b/tests/functional/u/unused/unused_private_member.py
@@ -66,3 +66,14 @@ class MyCls:
     @classmethod
     def get_class_var(cls):
         return cls.__class_var
+
+
+class Bla:
+    """Regression test for issue 4638"""
+
+    def __init__(self):
+        type(self).__a()
+
+    @classmethod
+    def __a(cls):
+        pass

--- a/tests/functional/u/unused/unused_private_member.py
+++ b/tests/functional/u/unused/unused_private_member.py
@@ -73,7 +73,17 @@ class Bla:
 
     def __init__(self):
         type(self).__a()
+        self.__b()
+        Bla.__c()
 
     @classmethod
     def __a(cls):
+        pass
+
+    @classmethod
+    def __b(cls):
+        pass
+
+    @classmethod
+    def __c(cls):
         pass

